Add initial database patterns

gridhead opened the pull request:
If you see some mistakes or discrepancies in the code, that is because that PR is still a work in progress. Pretty much how someone justifies the presence of bugs in their software just because they have not yet hit their 1.0 release.

Fixes #63

Let's hit ~our 1.0 release~ flesh out this PR together ;P

Signed-off-by: Akashdeep Dhar <akashdeep.dhar@gmail.com>

gridhead committed:
Add basic database model and framework

The model uses joint table inheritance between Node and its subclasses
and adds a declarative enum type and supporting code.

Additionally, provide a global session object which is usable from
different contexts (e.g. threads) and define how certain automatically
generated objects in the database are named.

Signed-off-by: Akashdeep Dhar <akashdeep.dhar@gmail.com>
Signed-off-by: Nils Philippsen <nils@redhat.com>
Co-authored-by: Nils Philippsen <nils@redhat.com>

gridhead commented on duffy/models/users.py:
```diff
@@ -0,0 +1,52 @@
+    flavour = Column(Enum, nullable=False)
+
+
+class PhysicalNode:
```
```suggestion
class PhysicalNode(Node):
```

gridhead commented on duffy/models/users.py:
```diff
@@ -0,0 +1,52 @@
+    comment = Column(Text, nullable=True)
+
+
+class VirtualNode:
```
```suggestion
class VirtualNode(Node):
```

nphilipp commented on duffy/models/users.py:
```diff
@@ -0,0 +1,52 @@
+class Node:
+    __tablename__ = "nodes"
+    id = Column(Integer, primary_key=True, nullable=False)
+    hostname = Column(Text, nullable=False)
```
```
    type = Column(NodeTypeEnum, nullable=False)
...
    __mapper_args__ = {
        "polymorphic_on": type,
    }
```

nphilipp commented on duffy/models/users.py:
```diff
@@ -0,0 +1,52 @@
+
+class Node:
+    __tablename__ = "nodes"
+    id = Column(Integer, primary_key=True, nullable=False)
```
```suggestion
    __mapper_args__ = {"polymorphic_on": "type"}
    id = Column(Integer, primary_key=True, nullable=False)
    type = Column(NodeType, nullable=False)
```

nphilipp commented on duffy/models/users.py:
```diff
@@ -0,0 +1,52 @@
+
+
+class VirtualNode:
+    __tablename__ = "virtualnodes"
```
```suggestion
    __tablename__ = "virtualnodes"
    __mapper_args__ = {"polymorphic_identity": NodeType.VIRTUAL}
```

nphilipp commented on duffy/models/users.py:
```diff
@@ -0,0 +1,52 @@
+
+
+class PhysicalNode:
+    __tablename__ == "physicalnodes"
```
```suggestion
    __tablename__ = "physicalnodes"
    __mapper_args__ = {"polymorphic_identity": NodeType.PHYSICAL}
```

nphilipp committed:
Configure isort line length

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Add tests for DB models and declarative enum type

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Update ER diagram

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Make test modules part of a proper `tests` package

While pytest doesn't require that test modules are in proper Python
packages, it has to modify `sys.path` in order to be able to load such
modules. This will fail if test modules have the same (base) file names.

See pytest documentation for more information:
https://docs.pytest.org/en/latest/explanation/pythonpath.html#standalone-test-modules-conftest-py-files

Signed-off-by: Nils Philippsen <nils@redhat.com>
